src/cli_utils.cc

```diff
@@ -78,70 +78,165 @@ Status LoadYAMLFile(const string& filename, YAML::Node &node) {
+    } catch (exception e) {
+        string err = e.what();
+        return Status::Failure("exception caught in yaml_verify_begin_doc_list: ", err);
     }
+}
+
+// In a YAML document built as a of document list, get the next document.
+//
+// Notes:
+// -  Catch any exceptions, and convert to bad status
+static Status yaml_get_next_document(std::istream &is,
+                                     YAML::Node &ans,
+                                     string& next_marker) {
+    try {
+        stringstream ss;
+
+        while (is.good() && !is.eof()) {
+            string line;
+            std::getline(is, line);
+
+            if (line.size() == 3) {
+                // check if we reached the end of this top level document
+                if (line == yaml_begin_doc ||
+                    line == yaml_end_doc_list) {
+                    // We have the entire document in memory. Convert to
+                    // a YAML node and return.
+                    next_marker = line;
+                    ans = std::move(YAML::Load(ss.str()));
+                    return Status::OK();
+                }
+            }
+
+            ss.write(line.c_str(), line.size());
+            ss.write("\n", 1);
+        }
+
+        return Status::Invalid("premature end of document, did not find End-of-Doc/End-of-File marker");
+    } catch (exception e) {
+        string err = e.what();
+        return Status::Failure("exception caught in yaml_get_next_document: ", err);
+    }
+}
+
+
+Status discovered_alleles_of_yaml_stream(std::istream &is,
+                                         std::vector<std::pair<std::string,size_t> > &contigs,
+                                         discovered_alleles &dsals) {
+    Status s;
+    string next_marker;
+    YAML::Node doc;
 
     contigs.clear();
     dsals.clear();
+    S(yaml_verify_begin_doc_list(is));
 
-    // read contigs
-    auto n_contigs = yaml["contigs"];
-    if (!n_contigs.IsSequence()) {
-        return Status::Invalid("contigs should be a yaml sequence");
-    }
-    for (auto p = n_contigs.begin(); p != n_contigs.end(); ++p) {
-        const std::string name = (*p)["name"].as<std::string>();
-        size_t size = (*p)["size"].as<size_t>();
-        contigs.push_back(make_pair(name, size));
-    }
+    // The first top-level document is the contigs
+    S(yaml_get_next_document(is, doc, next_marker));
+    S(contigs_of_yaml(doc, contigs));
+
+    // All other documents are discovered-alleles
+    while (next_marker != yaml_end_doc_list) {
+        S(yaml_get_next_document(is, doc, next_marker));
 
-    // read ranges and alleles
-    auto n_ranges = yaml["alleles"];
-    s = discovered_alleles_of_yaml(n_ranges, contigs, dsals);
-    if (s.bad()) return s;
+        allele allele(range(-1,-1,-1), "A");
+        discovered_allele_info ainfo;
 
+        S(one_discovered_allele_of_yaml(doc, contigs, allele, ainfo));
+        dsals[allele] = ainfo;
+    }
+    if (next_marker != yaml_end_doc_list)
```

Comment:
I think this if statement cannot be true at this point based on the previous loop